File: events/year_2017/day_17/day_17.rs
aoc::main!();

fn preprocessing(puzzle_input: &str) -> usize {
    puzzle_input
    .parse()
    .unwrap()
}

fn part_1(steps: usize) -> usize {
    *(0..=2017)
    .fold((0, vec![0]), |(mut pos, mut state), entry| {
        pos = 1 + (pos + steps) % (entry + 1);
        state.insert(pos, entry + 1);
        (pos, state)})
    .1
    .iter()
    .skip_while(|n| **n != 2017)
    .nth(1)
    .unwrap()}

fn part_2(steps: usize) -> usize {
    (0..=50_000_000)
    .fold((0, 0), |(value, pos), entry| 
        (
            if pos == 1 {entry} else {value}
        ,
            1 + (pos + steps) % (entry + 1)
        ))
    .0
}

File: events/year_2017/day_17/day_17.py
"""
Advent of Code - Year 2017 - Day 17
https://adventofcode.com/2017/day/17
"""

from collections.abc import Iterator


def preprocessing(puzzle_input: str) -> int:
    """
    Converts the puzzle input string to an integer.
    """
    return int(puzzle_input)


def solver(n_steps: int) -> Iterator[int]:
    """
    Simulates a circular buffer spinlock and computes values at specified positions.

    The function iterates through a simulation of 50 million steps, where each step involves
    calculating a new position and inserting a value. It tracks:
    - The value at position 1 after 50 million insertions
    - The value immediately after 2017 in the state list after 2017 insertions

    Args:
        n_steps (int): Number of steps to move forward in each iteration

    Yields:
        int: First, the value that comes after 2017 in the buffer after 2017 insertions
        int: Second, the value at position 1 after 50 million insertions
    """
    value = 0
    pos = 0
    state = [0]
    for i in range(50_000_000):
        pos = 1 + (pos + n_steps) % (i + 1)
        if pos == 1:
            value = i + 1
        if i < 2017:
            state.insert(pos, i + 1)
        if i == 2017:
            yield state[state.index(2017) + 1]
    yield value
